crates/slot-arithmetic/src/lib.rs

```diff
@@ -183,9 +184,66 @@ impl Step for Epoch {
     }
 }
 
+#[derive(Clone, Debug, Copy, PartialEq, PartialOrd, Ord, Eq, Serialize, Deserialize, Default)]
+#[repr(transparent)]
+pub struct TimeMs(u64);
+
+impl From<u64> for TimeMs {
+    fn from(epoch: u64) -> TimeMs {
+        TimeMs(epoch)
+    }
+}
+
+impl From<TimeMs> for u64 {
+    fn from(time_ms: TimeMs) -> u64 {
+        time_ms.0
+    }
+}
+
+impl Add<u64> for TimeMs {
+    type Output = Self;
+
+    fn add(self, rhs: u64) -> Self::Output {
+        TimeMs(self.0 + rhs)
+    }
+}
+
+impl<C> Encode<C> for TimeMs {
+    fn encode<W: minicbor::encode::Write>(
+        &self,
+        e: &mut minicbor::Encoder<W>,
+        ctx: &mut C,
+    ) -> Result<(), minicbor::encode::Error<W::Error>> {
+        self.0.encode(e, ctx)
+    }
+}
+
+impl<'b, C> Decode<'b, C> for TimeMs {
+    fn decode(d: &mut Decoder<'b>, _ctx: &mut C) -> Result<Self, minicbor::decode::Error> {
+        d.u64()
+            .or_else(|_| {
+                let bytes: &[u8] = d.bytes()?;
+                let num = BigUint::from_bytes_be(bytes);
+                let num_ms = num / 1_000_000_000u64;
```

Comment:
Are we fine cropping to millis? Why not keep the full resolution?

Reply:
I would argue the value does not even need to have a finer grained resolution than _seconds_ but yes, I will add some comment.